File: 2025/October-31-2025/topk.py
from typing import List

nums = [1,2,2,3,3,3]
k = 2


def topK(nums: List[int], k: int) -> List[int]: 
  n = len(nums)
  count = {}
  freq = [[] for i in range(0, len(nums) + 1)]

  for num in nums: 
    count[num] = count.get(num, 0) + 1
  for num in count: 
    freq[count[num]].append(num)
  res = []

  for i in range(len(nums), -1, -1): 
    for f in freq[i]:
      res.append(f)
      if len(res) == k: 
        res.sort()
        return res
  return res


print(topK(nums, k))

File: 2025/October-17-2025/topk.js
// Top K Frequent Elements
// Solved 
// Given an integer array nums and an integer k, return the k most frequent elements within the array.

// The test cases are generated such that the answer is always unique.

// You may return the output in any order.

// Example 1:

// Input: nums = [1,2,2,3,3,3], k = 2

// Output: [2,3]

nums = [1,2,2,3,3,3], k = 2

const topK = (nums, k) => { 
  const count = {}
  const freq = Array.from({length: nums.length + 1}, () => [])

  for(const num of nums){ 
    count[num] = (count[num] || 0) + 1
  }

  for(const num in count){ 
    freq[count[num]].push(parseInt(num))
  }

  console.log(freq)

  const res = []

  for(let i = freq.length - 1; i > 0; i--){ 
    for(const n of freq[i]){
      res.push(n)
      if(res.length === k){
        return res
      }
    }
  }
}

console.log(topK(nums, k))